nlpcore/src/latin/dictionary.cppm

```diff
@@ -72,38 +75,45 @@ export class LatinDictionary : public Dictionary {
     LatinDictionary& operator=(LatinDictionary&&) = default;
 
     inline LatinTrieNode* insertNgram(std::span<const fl::str::UniString> ngram) noexcept {
-        return algorithms::insertNgram(data_.get(), dict_id_, ngram);
+        std::scoped_lock<std::shared_mutex> lock(data_->second);
+        return algorithms::insertNgram(&(data_->first), dict_id_, ngram);
     }
 
-    inline void forEachEntry(algorithms::EntryAction& action) {
-        algorithms::forEachEntry(data_.get(), dict_id_, action);
+    inline void forEachEntryReadSafe(algorithms::EntryAction& action) {
+        std::shared_lock<std::shared_mutex> lock(data_->second);
+        algorithms::forEachEntry(&(data_->first), dict_id_, action);
     }
```

Comment:
Because scoped_lock is non-shared lock (write lock) while shared_lock is shared (read lock). I'd use scoped_shared_lock if there is one, but there isn't.